types/node/util.d.ts

```diff
@@ -1105,6 +1105,197 @@ declare module 'util' {
+    They differ only in the final item of the if-else chain.
+    */
+    type IfDefaultsTrue<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfTrue;
+
+    type IfDefaultsFalse<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfFalse;
+
+    type ExtractOptionValue<T extends ParseArgsConfig, O extends ParseArgsOptionConfig> = IfDefaultsTrue<
+        T['strict'],
+        O['type'] extends 'string' ? string : O['type'] extends 'boolean' ? boolean : string | boolean,
+        string | boolean
+    >;
+
+    type ParsedValues<T extends ParseArgsConfig> = (T['options'] extends ParseArgsOptionsConfig
+        ? {
+              -readonly [LongOption in keyof T['options']]: IfDefaultsFalse<
+                  T['options'][LongOption]['multiple'],
+                  undefined | Array<ExtractOptionValue<T, T['options'][LongOption]>>,
+                  undefined | ExtractOptionValue<T, T['options'][LongOption]>
+              >;
+          }
+        : {}) &
+        IfDefaultsTrue<T['strict'], unknown, { [longOption: string]: undefined | string | boolean }>;
+
+    type ParsedPositionals<T extends ParseArgsConfig> = IfDefaultsTrue<
+        T['strict'],
+        IfDefaultsFalse<T['allowPositionals'], string[], []>,
+        IfDefaultsTrue<T['allowPositionals'], string[], []>
+    >;
+
+    type PreciseTokenForOptions<
+        T extends ParseArgsConfig,
+        K extends keyof T['options'] & string,
+        O extends ParseArgsOptionConfig,
+    > = O['type'] extends 'string'
+        ? {
+              kind: 'option';
+              index: number;
+              name: K;
+              rawName: string;
+              value: string;
+              inlineValue: boolean;
+          }
+        : O['type'] extends 'boolean'
+        ? {
+              kind: 'option';
+              index: number;
+              name: K;
+              rawName: string;
+              value: undefined;
+              inlineValue: undefined;
+          }
+        : OptionToken & { name: K };
+
+    type TokenForOptions<
+        T extends ParseArgsConfig,
+        K extends keyof T['options'] = keyof T['options'],
+    > = K extends unknown
+        ? T['options'] extends ParseArgsOptionsConfig
+            ? PreciseTokenForOptions<T, K & string, T['options'][K]>
+            : OptionToken
+        : never;
+
+    type ParsedOptionToken<T extends ParseArgsConfig> = IfDefaultsTrue<T['strict'], TokenForOptions<T>, OptionToken>;
+
+    type ParsedPositionalToken<T extends ParseArgsConfig> = IfDefaultsTrue<
+        T['strict'],
+        IfDefaultsFalse<T['allowPositionals'], { kind: 'positional'; index: number; value: string }, never>,
```

Comment:
While I can definitely understand the need for type safety around `values`, this level of complexity seems overkill for `tokens`. How often will `tokens` be used compared to `values`, and does that justify the maintenance burden incurred for these types?

Reply:
I expect `tokens` to be used reasonable often, though certainly much less often than `values`. It's pretty handy for doing more sophisticated things.

What maintenance burden are you concerned about? I don't expect this part of the API to change much, if ever, and am happy to commit to maintain the types here if it ever does (just as I would be doing if this package were published on npm instead of as part of node core).